### PR TITLE
Add Media Sanitize Protocol for Nvme devices.

### PR DESCRIPTION
**Pull request by apop5**
# Description

Update IndustryStandard/Nvme.h to include additional fields from the 1.4c specification.

Add media sanitize protocol to the NvmeExpressDxe driver.

- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [X] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested
N/A

## Integration Instructions
N/A

### CODE REVIEW

**Comment by lgao4** [comment]
This is a new protocol in MdeModulePkg. What's its usage model?

**Comment by apop5** [comment]
This is a protocol makes available the interfaces from https://nvlpubs.nist.gov/nistpubs/specialpublications/nist.sp.800-88r1.pdf

The expected usage model is that an external driver will be able to make use of the protocol to perform the NIST compatible sanitization of the media.

File: MdePkg/Include/IndustryStandard/Nvme.h [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
   (C) Copyright 2016 Hewlett Packard Enterprise Development LP<BR>
   Copyright (c) 2017 - 2023, Intel Corporation. All rights reserved.<BR>
+  Copyright (c) Microsoft Corporation.
```
Comment (leiflindholm) [inline MdePkg/Include/IndustryStandard/Nvme.h:6]: Could you add the <BR> at the end for doxygen, please?

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c [hunk 2]
```diff
@@ -3,6 +3,7 @@
   NVM Express specification.
 
   Copyright (c) 2013 - 2017, Intel Corporation. All rights reserved.<BR>
+  Copyright (c) Microsoft Corporation.
```
Comment (leiflindholm) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:6]: Could you add the `<BR>` at the end for doxygen, please?

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
   (C) Copyright 2016 Hewlett Packard Enterprise Development LP<BR>
   Copyright (c) 2013 - 2019, Intel Corporation. All rights reserved.<BR>
+  Copyright (c) Microsoft Corporation.
```
Comment (leiflindholm) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:7]: Could you add the `<BR>` at the end for doxygen, please?

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h [hunk 4]
```diff
@@ -0,0 +1,191 @@
+/** @file
+  Header file for MEDIA_SANITIZE_PROTOCOL interface.
+
+  Copyright (c) Microsoft Corporation.
```
Comment (leiflindholm) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:4]: Could you add a `<BR>` at the end for doxygen, please?

File: MdeModulePkg/Include/Protocol/MediaSanitize.h [hunk 5]
```diff
@@ -0,0 +1,173 @@
+/** @file
+  This file defines the Media Sanitize Protocol.
+
+  Copyright (c) Microsoft Corporation.
```
Comment (leiflindholm) [inline MdeModulePkg/Include/Protocol/MediaSanitize.h:4]: Could you add a `<BR>` at the end for doxygen, please?

File: MdePkg/Include/IndustryStandard/Nvme.h (hunk 1 above)
Comment (apop5) [reply MdePkg/Include/IndustryStandard/Nvme.h]: Updated in the latest push. Thanks for reviewing.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c (hunk 2 above)
Comment (apop5) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c]: Updated in the latest push. Thanks for reviewing.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h (hunk 3 above)
Comment (apop5) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h]: Updated in the latest push. Thanks for reviewing.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressDxe.inf [hunk 6]
```diff
@@ -5,7 +5,7 @@
 #  NVM Express specification.
 #
 #  Copyright (c) 2013 - 2019, Intel Corporation. All rights reserved.<BR>
-#
+#  Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (apop5) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressDxe.inf]: Updated in the latest push. Thanks for reviewing.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h (hunk 4 above)
Comment (apop5) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h]: Updated in the latest push. Thanks for reviewing.

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTestHost.inf [hunk 7]
```diff
@@ -0,0 +1,37 @@
+## @file
+# Unit tests for MEDIA_SANITIZE_PROTOCOL and mapping to NVM Express native commands (Sanitize and FormatNVM)
+#
+# Copyright (C) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (apop5) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTestHost.inf]: Updated in the latest push. Thanks for reviewing.

File: MdeModulePkg/Include/Protocol/MediaSanitize.h (hunk 5 above)
Comment (apop5) [reply MdeModulePkg/Include/Protocol/MediaSanitize.h]: Updated in the latest push. Thanks for reviewing.

**Review by lgao4 — APPROVED** [review]

**Comment by lgao4** [comment]
I have no comments for this patch.

**Comment by apop5** [comment]
@lgao4 @mdkinney @leiflindholm 

Are there any other concerns for this PR?

File: MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c [hunk 8]
```diff
@@ -181,6 +182,26 @@ EnumerateNvmeDevNamespace (
...
+      sizeof (Device->Controller->ControllerData->Sanicap)
+      );
+
+    CopyMem (
+      &(Device->MediaSanitize.SanitizeCapabilities),
+      &(Device->Controller->ControllerData->Sanicap),
+      sizeof (Device->MediaSanitize.SanitizeCapabilities)
+      );
+
     //
```
Comment (lgao4) [inline MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:205]: Here MediaSanitize is initialized. But, this protocol is not installed. Is this by purpose not to install it?

Comment (apop5) [reply MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:205]: Thank you for catching that. It looks like it was an oversight in creating the pr. 

It has been updated in the latest push.

**Pull request merged**